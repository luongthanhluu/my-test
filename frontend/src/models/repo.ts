import { Card } from './card'

export interface RepoItem {
    id: string
    title: string
    cards: Card[]
}

export interface Repo {
    id?: string
    name: string
    lists?: RepoItem[]
}

export enum ListStatus {
    Open = 'Open',
    Confirmed = 'Confirmed',
    FalsePositive = 'False Positive',
    Fixed = 'Fixed',
}
