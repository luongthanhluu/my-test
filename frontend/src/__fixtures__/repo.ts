import { Repo } from 'models/repo'

export const repoMock: Repo[] = [
    {
        name: 'Repo 1',
        id: '1',
        lists: [
            { title: 'Open', cards: [], id: 'a' },
            {
                title: 'Confirmed',
                cards: [
                    { text: 'Vulnerability 3', id: 'y' },
                    { text: 'Vulnerability 4', id: 'z' },
                ],
                id: 'b',
            },
            {
                title: 'False Positive',
                cards: [
                    { text: 'Vulnerability 5', id: 'm' },
                    { text: 'Vulnerability 6', id: 'n' },
                ],
                id: 'c',
            },
            {
                title: 'Fixed',
                cards: [
                    { text: 'Vulnerability 7', id: 'o' },
                    { text: 'Vulnerability 8', id: 'p' },
                    {
                        text: 'Vulnerability 1',
                        id: 'w',
                        activities: [
                            {
                                createdAt: '2021-11-28T11:24:05.600Z',
                                description: 'moved from Open to Confirmed',
                            },
                            {
                                createdAt: '2021-11-28T11:24:24.903Z',
                                description: 'moved from Confirmed to Fixed',
                            },
                        ],
                    },
                ],
                id: 'd',
            },
        ],
    },
]
