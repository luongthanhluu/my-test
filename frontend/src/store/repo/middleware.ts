import { createAsyncThunk } from '@reduxjs/toolkit'
import { Repo } from 'models/repo'
import {
    getListRepoApi,
    editRepoApi,
    addRepoApi,
    deleteRepoApi,
} from './../../services/repoApi'

export const repoList = createAsyncThunk('repo/repoList', async () => {
    const response = await getListRepoApi()
    return response
})

export const updateRepo = createAsyncThunk(
    'repo/updateRepo',
    async (repo: Repo) => {
        const response = await editRepoApi(repo)
        return response
    }
)

export const addRepo = createAsyncThunk('repo/addRepo', async (repo: Repo) => {
    const response = await addRepoApi(repo)
    return response
})

export const deleteRepo = createAsyncThunk(
    'repo/deleteRepo',
    async (id: string) => {
        const response = await deleteRepoApi(id)
        return response
    }
)
