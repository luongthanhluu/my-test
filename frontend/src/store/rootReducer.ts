import { combineReducers } from '@reduxjs/toolkit'

import { repoSlice } from './repo'
import { boardSlice } from './board'

export const rootReducer = combineReducers({
    repo: repoSlice.reducer,
    board: boardSlice.reducer,
})

export type RootState = ReturnType<typeof rootReducer>
