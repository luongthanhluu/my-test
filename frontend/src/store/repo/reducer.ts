import { createSlice, current, PayloadAction } from '@reduxjs/toolkit'
import {
    ERROR_ADD_REPO_MESSAGE,
    ERROR_DELETE_REPO_MESSAGE,
    ERROR_UPDATE_REPO_MESSAGE,
    ERROR_LOAD_REPO_MESSAGE,
} from 'constants/message'
import { Repo } from 'models/repo'
import { getIndexRepoById } from 'utils/repoHelper'
import { repoList, updateRepo, addRepo, deleteRepo } from './middleware'

export interface RepoState {
    list: {
        loading: boolean
        data: Repo[]
        errorMessage: string
    }
}

const initialState: RepoState = {
    list: {
        loading: false,
        data: [],
        errorMessage: '',
    },
}

// Ref: https://redux-toolkit.js.org/tutorials/basic-tutorial
// Slice can create actions and reducer together
export const repoSlice = createSlice({
    name: 'repo',
    initialState: initialState,
    reducers: {},
    extraReducers: {
        //get list repo
        [repoList.pending.toString()]: (state) => {
            state.list.loading = true
        },
        [repoList.fulfilled.toString()]: (
            state,
            { payload }: PayloadAction<Repo[]>
        ) => {
            state.list.loading = false
            state.list.data = payload
        },
        [repoList.rejected.toString()]: (state) => {
            state.list.loading = false
            state.list.errorMessage = ERROR_LOAD_REPO_MESSAGE
        },

        //update repo name
        [updateRepo.pending.toString()]: (state) => {
            state.list.loading = true
        },
        [updateRepo.fulfilled.toString()]: (
            state,
            { payload }: PayloadAction<Repo>
        ) => {
            state.list.loading = false
            const listRepo = current(state.list.data)
            const index = getIndexRepoById(listRepo, payload?.id || '')
            if (index > -1) {
                state.list.data[index] = payload
            }
        },
        [updateRepo.rejected.toString()]: (state) => {
            state.list.loading = false
            state.list.errorMessage = ERROR_UPDATE_REPO_MESSAGE
        },

        // add new repo
        [addRepo.pending.toString()]: (state) => {
            state.list.loading = true
        },
        [addRepo.fulfilled.toString()]: (
            state,
            { payload }: PayloadAction<Repo>
        ) => {
            state.list.loading = false
            state.list.data = [...state.list.data, payload]
        },
        [addRepo.rejected.toString()]: (state) => {
            state.list.loading = false
            state.list.errorMessage = ERROR_ADD_REPO_MESSAGE
        },

        // delete new repo
        [deleteRepo.pending.toString()]: (state) => {
            state.list.loading = true
            state.list.errorMessage = ''
        },
        [deleteRepo.fulfilled.toString()]: (
            state,
            { payload }: PayloadAction<string>
        ) => {
            state.list.loading = false
            const listRepo = current(state.list.data)
            const index = getIndexRepoById(listRepo, payload || '')
            if (index > -1) {
                const newList = [...listRepo]
                newList.splice(index, 1)
                state.list.data = newList
            }
        },
        [deleteRepo.rejected.toString()]: (state) => {
            state.list.loading = false
            state.list.errorMessage = ERROR_DELETE_REPO_MESSAGE
        },
    },
})
