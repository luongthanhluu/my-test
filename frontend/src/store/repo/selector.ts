import { RootState } from '../rootReducer'

export const getListRepoData = (state: RootState) => state.repo?.list
