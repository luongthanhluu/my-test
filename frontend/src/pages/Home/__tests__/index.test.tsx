import React from 'react'
import { render } from 'enzyme'
import { Provider } from 'react-redux'

import { store } from 'store'
import { HomeComponent } from '../index'
import { repoMock } from '../../../__fixtures__/repo'

jest.mock('react-router-dom', () => ({
    ...(jest.requireActual('react-router-dom') as any),
    useNavigate: jest.fn(),
}))

test('should match snapshot with data', () => {
    const wrapper = render(
        <Provider store={store}>
            <HomeComponent loading={false} data={repoMock} errorMessage="" />
        </Provider>
    )
    expect(wrapper).toMatchSnapshot()
})

test('should match snapshot with error message', () => {
    const wrapper = render(
        <Provider store={store}>
            <HomeComponent
                loading={false}
                data={repoMock}
                errorMessage="error message"
            />
        </Provider>
    )
    expect(wrapper).toMatchSnapshot()
})

test('should match snapshot with loading', () => {
    const wrapper = render(
        <Provider store={store}>
            <HomeComponent
                loading={false}
                data={repoMock}
                errorMessage="error message"
            />
        </Provider>
    )
    expect(wrapper).toMatchSnapshot()
})
