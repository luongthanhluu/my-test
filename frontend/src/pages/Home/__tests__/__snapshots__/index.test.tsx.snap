// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should match snapshot with data 1`] = `
<div>
  <header
    class="MuiPaper-root MuiPaper-elevation MuiPaper-elevation4 MuiAppBar-root MuiAppBar-colorPrimary MuiAppBar-positionStatic css-hip9hq-MuiPaper-root-MuiAppBar-root"
  >
    <div
      class="MuiToolbar-root MuiToolbar-gutters MuiToolbar-dense css-10ltm04-MuiToolbar-root"
    >
      <div
        class="MuiTypography-root MuiTypography-h6 css-1qp5wg9-MuiTypography-root"
      >
        Home
      </div>
    </div>
  </header>
  <div
    class="MuiContainer-root MuiContainer-maxWidthLg css-13kvln1-MuiContainer-root"
  >
    <nav
      class="MuiList-root MuiList-padding css-h4y409-MuiList-root"
    >
      <li
        class="MuiListItem-root MuiListItem-gutters css-seb0jc-MuiListItem-root"
      >
        <div
          class="MuiListItemButton-root MuiListItemButton-gutters MuiButtonBase-root css-cvhtoe-MuiButtonBase-root-MuiListItemButton-root"
          role="button"
          tabindex="0"
        >
          <div
            class="MuiListItemText-root css-tlelie-MuiListItemText-root"
          >
            <span
              class="MuiTypography-root MuiTypography-body1 MuiListItemText-primary css-10hburv-MuiTypography-root"
            >
              Repo 1
            </span>
          </div>
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </div>
        <div
          class="MuiListItemSecondaryAction-root css-518kzi-MuiListItemSecondaryAction-root"
        >
          <button
            class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-78trlr-MuiButtonBase-root-MuiIconButton-root"
            tabindex="0"
            type="button"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-colorPrimary MuiSvgIcon-fontSizeMedium css-ci5apu-MuiSvgIcon-root"
              data-testid="EditIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M3 17.25V21h3.75L17.81 9.94l-3.75-3.75L3 17.25zM20.71 7.04c.39-.39.39-1.02 0-1.41l-2.34-2.34a.9959.9959 0 0 0-1.41 0l-1.83 1.83 3.75 3.75 1.83-1.83z"
              />
            </svg>
          </button>
        </div>
      </li>
    </nav>
    <div
      class="MuiBox-root css-15z49vq"
    >
      <button
        class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-78trlr-MuiButtonBase-root-MuiIconButton-root"
        tabindex="0"
        type="button"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-colorSecondary MuiSvgIcon-fontSizeLarge css-1nryff-MuiSvgIcon-root"
          data-testid="AddCircleIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm5 11h-4v4h-2v-4H7v-2h4V7h2v4h4v2z"
          />
        </svg>
      </button>
    </div>
  </div>
</div>
`;

exports[`should match snapshot with error message 1`] = `
<div>
  <header
    class="MuiPaper-root MuiPaper-elevation MuiPaper-elevation4 MuiAppBar-root MuiAppBar-colorPrimary MuiAppBar-positionStatic css-hip9hq-MuiPaper-root-MuiAppBar-root"
  >
    <div
      class="MuiToolbar-root MuiToolbar-gutters MuiToolbar-dense css-10ltm04-MuiToolbar-root"
    >
      <div
        class="MuiTypography-root MuiTypography-h6 css-1qp5wg9-MuiTypography-root"
      >
        Home
      </div>
    </div>
  </header>
  <div
    class="MuiContainer-root MuiContainer-maxWidthLg css-13kvln1-MuiContainer-root"
  >
    <div
      class="MuiPaper-root MuiPaper-elevation MuiPaper-elevation0 MuiAlert-root MuiAlert-standardError MuiAlert-standard css-145dpyk-MuiPaper-root-MuiAlert-root"
      role="alert"
    >
      <div
        class="MuiAlert-icon css-1ytlwq5-MuiAlert-icon"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit css-1vooibu-MuiSvgIcon-root"
          data-testid="ErrorOutlineIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M11 15h2v2h-2zm0-8h2v6h-2zm.99-5C6.47 2 2 6.48 2 12s4.47 10 9.99 10C17.52 22 22 17.52 22 12S17.52 2 11.99 2zM12 20c-4.42 0-8-3.58-8-8s3.58-8 8-8 8 3.58 8 8-3.58 8-8 8z"
          />
        </svg>
      </div>
      <div
        class="MuiAlert-message css-acap47-MuiAlert-message"
      >
        error message
      </div>
    </div>
    <nav
      class="MuiList-root MuiList-padding css-h4y409-MuiList-root"
    >
      <li
        class="MuiListItem-root MuiListItem-gutters css-seb0jc-MuiListItem-root"
      >
        <div
          class="MuiListItemButton-root MuiListItemButton-gutters MuiButtonBase-root css-cvhtoe-MuiButtonBase-root-MuiListItemButton-root"
          role="button"
          tabindex="0"
        >
          <div
            class="MuiListItemText-root css-tlelie-MuiListItemText-root"
          >
            <span
              class="MuiTypography-root MuiTypography-body1 MuiListItemText-primary css-10hburv-MuiTypography-root"
            >
              Repo 1
            </span>
          </div>
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </div>
        <div
          class="MuiListItemSecondaryAction-root css-518kzi-MuiListItemSecondaryAction-root"
        >
          <button
            class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-78trlr-MuiButtonBase-root-MuiIconButton-root"
            tabindex="0"
            type="button"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-colorPrimary MuiSvgIcon-fontSizeMedium css-ci5apu-MuiSvgIcon-root"
              data-testid="EditIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M3 17.25V21h3.75L17.81 9.94l-3.75-3.75L3 17.25zM20.71 7.04c.39-.39.39-1.02 0-1.41l-2.34-2.34a.9959.9959 0 0 0-1.41 0l-1.83 1.83 3.75 3.75 1.83-1.83z"
              />
            </svg>
          </button>
        </div>
      </li>
    </nav>
    <div
      class="MuiBox-root css-15z49vq"
    >
      <button
        class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-78trlr-MuiButtonBase-root-MuiIconButton-root"
        tabindex="0"
        type="button"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-colorSecondary MuiSvgIcon-fontSizeLarge css-1nryff-MuiSvgIcon-root"
          data-testid="AddCircleIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm5 11h-4v4h-2v-4H7v-2h4V7h2v4h4v2z"
          />
        </svg>
      </button>
    </div>
  </div>
</div>
`;

exports[`should match snapshot with loading 1`] = `
<div>
  <header
    class="MuiPaper-root MuiPaper-elevation MuiPaper-elevation4 MuiAppBar-root MuiAppBar-colorPrimary MuiAppBar-positionStatic css-hip9hq-MuiPaper-root-MuiAppBar-root"
  >
    <div
      class="MuiToolbar-root MuiToolbar-gutters MuiToolbar-dense css-10ltm04-MuiToolbar-root"
    >
      <div
        class="MuiTypography-root MuiTypography-h6 css-1qp5wg9-MuiTypography-root"
      >
        Home
      </div>
    </div>
  </header>
  <div
    class="MuiContainer-root MuiContainer-maxWidthLg css-13kvln1-MuiContainer-root"
  >
    <div
      class="MuiPaper-root MuiPaper-elevation MuiPaper-elevation0 MuiAlert-root MuiAlert-standardError MuiAlert-standard css-145dpyk-MuiPaper-root-MuiAlert-root"
      role="alert"
    >
      <div
        class="MuiAlert-icon css-1ytlwq5-MuiAlert-icon"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit css-1vooibu-MuiSvgIcon-root"
          data-testid="ErrorOutlineIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M11 15h2v2h-2zm0-8h2v6h-2zm.99-5C6.47 2 2 6.48 2 12s4.47 10 9.99 10C17.52 22 22 17.52 22 12S17.52 2 11.99 2zM12 20c-4.42 0-8-3.58-8-8s3.58-8 8-8 8 3.58 8 8-3.58 8-8 8z"
          />
        </svg>
      </div>
      <div
        class="MuiAlert-message css-acap47-MuiAlert-message"
      >
        error message
      </div>
    </div>
    <nav
      class="MuiList-root MuiList-padding css-h4y409-MuiList-root"
    >
      <li
        class="MuiListItem-root MuiListItem-gutters css-seb0jc-MuiListItem-root"
      >
        <div
          class="MuiListItemButton-root MuiListItemButton-gutters MuiButtonBase-root css-cvhtoe-MuiButtonBase-root-MuiListItemButton-root"
          role="button"
          tabindex="0"
        >
          <div
            class="MuiListItemText-root css-tlelie-MuiListItemText-root"
          >
            <span
              class="MuiTypography-root MuiTypography-body1 MuiListItemText-primary css-10hburv-MuiTypography-root"
            >
              Repo 1
            </span>
          </div>
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </div>
        <div
          class="MuiListItemSecondaryAction-root css-518kzi-MuiListItemSecondaryAction-root"
        >
          <button
            class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-78trlr-MuiButtonBase-root-MuiIconButton-root"
            tabindex="0"
            type="button"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-colorPrimary MuiSvgIcon-fontSizeMedium css-ci5apu-MuiSvgIcon-root"
              data-testid="EditIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M3 17.25V21h3.75L17.81 9.94l-3.75-3.75L3 17.25zM20.71 7.04c.39-.39.39-1.02 0-1.41l-2.34-2.34a.9959.9959 0 0 0-1.41 0l-1.83 1.83 3.75 3.75 1.83-1.83z"
              />
            </svg>
          </button>
        </div>
      </li>
    </nav>
    <div
      class="MuiBox-root css-15z49vq"
    >
      <button
        class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-78trlr-MuiButtonBase-root-MuiIconButton-root"
        tabindex="0"
        type="button"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-colorSecondary MuiSvgIcon-fontSizeLarge css-1nryff-MuiSvgIcon-root"
          data-testid="AddCircleIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm5 11h-4v4h-2v-4H7v-2h4V7h2v4h4v2z"
          />
        </svg>
      </button>
    </div>
  </div>
</div>
`;
